### DISCUSSION TITLE
Interesting project with similarities to Crank: Whatsup

### DISCUSSION THREAD

**Discussion by canadaduane** [bikeshaving/crank#303] [General] [answered]
I thought this project was interesting in its design choices, which reminded me of Crank.JS:

https://whatsup.js.org/docs/intro

Not sure if it is valuable, or directly useful, but interesting nonetheless.

**Comment by unodgs** [comment]
Looking at lifecycle example here https://whatsup.js.org/docs/components/lifecycle I would say it's a copy of crank

**Comment by brainkim** [comment, marked as answer]
Converted to a discussion because there’s nothing actionable for me.

Whatsup definitely seems similar to Crank! It has the generator component pattern, which is, in my humble opinion, a great idea. But it uses observables for component state. The interesting thing about saying observables are the state of the component, is that the component is still stateful, and the generator scope can still have state in it which is not tracked in an observable, so the observable is just a fancy extra layer on top of the statefulness of generators.

```jsx
function* App() {
    const counter = observable(0)
    const increment = () => counter(counter() + 1)
    let i = 0;

    while (true) {
        // this is state which is not tracked by an observable
        i++;
        yield (
            <div>
                <p>You click {counter()} times</p>
                <button onClick={increment}>Click me</button>
            </div>
        )
    }
}
```

The advantage of using observables is that all state is formally declared. The disadvantage is that you have to use observables. But seriously, the other advantage is that observables can be threaded directly into elements to update only that prop:

```jsx

function App() {
    const className = observable<'red' | 'green'>('red')
    const handleClick = () => {
        const newValue = className() === 'red' ? 'green' : 'red'
        className(newValue)
    }

    while (true) {
        yield(
            <button className={className} onClick={handleClick}>
                Click me
            </button>
        )
    }
}

render(<App />)
```
This is performance thing I wish we could have but we can’t. Crank has to rely on wholesale element diffing, for better or worse.

One thing I’ve implemented in Crank which vindicates my decision to use props iterators, is to have the props iterator close when components are unmounted:
```jsx
function* Timer() {
  let seconds = 0;
  const interval = setInterval(() => this.refresh(() => seconds++), 1000);
  for ({} of this) {
    yield <div>Seconds: {seconds}</div>;
  }
  clearInterval(interval);
}
```

Relying on yields being returned and try/finally is an extra block that I don’t want to write, and `while (true)` is just annoying, scary, and gross looking. However, because whatsups is based on observables, you can’t have a props observable, because you can’t yield from inside the observable because it is a callback-based API. Ultimately I think generators and props iterators and manual `refresh()` wins out in terms of ergonomics.

Sorry for not replying sooner!

**Comment by iminside** [comment]
Hey, everyone!

Thanks for bringing this up — interesting discussion. Thanks @brainkim for the thoughtful write-up.

I'm the author of [Whatsup](https://whatsup.js.org/docs/intro). Thanks to @canadaduane and @unodgs for noting the similarities — the [lifecycle docs](https://whatsup.js.org/docs/components/lifecycle) really do look a lot like Crank's. Yes, that resemblance is real, but the projects didn't share a lineage — we converged on generators independently and made different bets underneath. I thought it might be useful to share how Whatsup actually came about, and respond to some of the points @brainkim raised.

---

### How it started (not as a JSX framework)

Whatsup began as experiments with state managers and generators, not as a UI library.

I'd worked with React + MobX for a long time and still use MobX in some projects — going back to the days when React components were classes and hooks didn't exist yet. But over time a few things got annoying: decorators like `@observable` going out of fashion, having to call `makeAutoObservable` explicitly in the constructor, and the pain of custom atoms. Whenever I needed a non-standard atom, it hurt.

A timer atom that starts on subscribe and tears down on unsubscribe is a good example. MobX gives you the tools, but the result is a lot of ceremony:

```ts
import { autorun, createAtom } from 'mobx'

const createTimer = () => {
    let counter = 0
    let interval: NodeJS.Timeout | null = null

    const tick = () => {
        counter++
        atom.reportChanged()
    }

    const onBecomeObserved = () => {
        interval = setInterval(tick, 1000)
    }

    const onBecomeUnobserved = () => {
        if (interval) {
            clearInterval(interval)
        }
        interval = null
        counter = 0
    }

    const atom = createAtom('Timer', onBecomeObserved, onBecomeUnobserved)

    return () => {
        atom.reportObserved()
        return counter
    }
}

const timer = createTimer()
const disposer = autorun(() => console.log(timer()))
disposer()
```

MobX gives you everything you need to build practically any atom you can imagine… but all of this machinery is just ugly.

Around the same time I was using generators in side projects, and it clicked: a generator naturally gives you a setup region, a loop for updates, and a place for teardown logic. I sketched something like:

```ts
const timer = computed(function* () {
    // setup
    const counter = observable(0)
    const interval = setInterval(() => counter(counter() + 1), 1000)

    try {
        while (true) {
            // update
            yield counter()
        }
    } finally {
        // teardown
        clearInterval(interval)
    }
})

const disposer = autorun(() => console.log(timer()))
disposer()
```

That looks great! The only thing that bothered me was `while (true)` — I wished JavaScript had a `loop {}` statement. It doesn't.

**After many iterations, Whatsup emerged** — a reactive state manager with a MobX-like API. The name symbolizes the constant dialogue between reactive entities: "What's up?" — a question about what changed, what's new in the system. `computed` atoms support two paradigms: the classical closure-based implementation and the more powerful one — generators with support for internal state.

Later, when the idea of creating a frontend framework arose, this state manager — [`@whatsup/core`](https://github.com/whatsup/whatsup/tree/master/packages/core) — became its central core.

---

### Mutators

While building the core I came up with **mutators**. A mutator is a simple function that receives the previous value and returns the next one. You pass a mutator into an `observable` atom to update its current value. In the timer example, a mutator fits naturally here:

```ts
// setup
const counter = observable(0)

const incrementMutator = (prev: number) => prev + 1 // mutator

const interval = setInterval(() => counter(incrementMutator), 1000)
```

Inside an atom, the new and old values are compared strictly with `===`. A mutator can return the previous value unchanged — in that case, propagation up the dependency tree stops at that atom. That makes mutators useful as filters.

For example, suppose we want a timer that only fires on even numbers:

```ts
const evenTimer = computed(() => {
    const next = timer()
    return (prev: number) => (next % 2 !== 0 ? prev : next)
})
```

The computed atom returns a mutator function with access to the atom's previous value. If the new value isn't even, the mutator returns the previous value and the update chain stops. Otherwise it returns the new value and the update propagates through the dependency tree.

In the console:

```ts
autorun(() => console.log(evenTimer()))

//> 2
//> 4
//> 6
```

What started as a small idea turned out to be a surprisingly powerful feature.

---

### Filters and comparers

The simple properties of mutators led to **filters** and **comparers** — convenient wrappers around the same idea.

Filters pass or block specific values:

```ts
const even = filter((next: number) => next % 2 === 0)

const evenTimer = computed(() => even(timer()))

autorun(() => console.log(evenTimer()))
//> 0
//> 2
//> 4
```

Comparers stop updates from propagating through the dependency tree when nothing meaningful changed:

```ts
const value = observable(0)

const equalsValueObj = comparer(
    (next, prev) =>
        next &&
        prev &&
        typeof next === 'object' &&
        typeof prev === 'object' &&
        Object.keys(next).length === 1 &&
        Object.keys(prev).length === 1 &&
        next.value === prev.value
)

const obj = computed(() => equalsValueObj({ value: value() }))
```

In MobX you'd do roughly:

```ts
const obj = computed(() => ({ value: value() }), { equals: equalsValueObj })
```

---

### From state manager to JSX

This is where Whatsup grew from a state manager into a full frontend framework with generator-based JSX. To be clear: the JSX renderer is just a layer on top. The reactive core — [`@whatsup/core`](https://github.com/whatsup/whatsup/tree/master/packages/core) with its atoms, mutators, and computed values — stays at the center. Components simply use that core to manage the DOM.

One day I asked: what if a `computed` atom's value is a DOM node? What if it updates that node through the mutator mechanism? The update instruction would be a virtual JSX node — compare old and new, apply the diff, return the result.

Every DOM element can be wrapped in a reactive atom. Full components can be built on generators. Every element, every component is an ordinary `computed` atom in one large dependency tree with a single root — from primitive data nodes (like `observable(0)`) all the way up to the root DOM node.

Generators let us avoid hooks entirely. The generator scope stays alive from render to render. Inside it you declare the atoms your render depends on, plus **reactions** — the analogue of effects. You skip the overhead of re-initializing hooks, comparing dependency arrays, and so on. That felt like a big win.

That's [`packages/jsx`](https://github.com/whatsup/whatsup/tree/master/packages/jsx) and the [Babel plugin to transform JSX](https://github.com/whatsup/whatsup/tree/master/packages/babel-plugin-transform-jsx).

I discovered Crank around this time and felt like I was on the right track. Yes — Whatsup and Crank are very similar. But that similarity is mostly surface-level, driven largely by generator syntax. @brainkim's work on Crank is excellent; I learned a huge amount from reading his source code, and it was genuinely useful while building my own renderer.

---

### Lifecycle: similar surface, different internals

The lifecycle shape in our docs is deliberately close to what generators suggest:

```tsx
function* App(props) {
    // componentWillMount
    try {
        // run render cycle
        while (true) {
            const newProps = yield <div>Hello</div>

            // componentWillReceiveProps

            props = newProps
        }
    } catch (e) {
        // componentDidCatch
    } finally {
        // componentDidUnmount
    }
}
```

Unlike Crank, I didn't use a props iterator. I chose to receive new props as the return value of `yield`, and to lean on native JavaScript control flow as much as possible.

The most contentious part is `while (true)`. I always hoped to improve it. But in pure JavaScript I never found a truly satisfying alternative.

I kept the `catch` and `finally` blocks and used them for what they're natively meant for. `catch` catches render errors in the component and all nested children — a full analogue of React's `ErrorBoundary`. `finally` always runs on unmount and cleanly separates teardown from the rest of the component.

Whatsup supports context, though honestly I'm not thrilled with how it's implemented today. On top of context there's a custom event system that bubbles through the context tree and lets you observe events in child components.

Because every DOM node is its own atom, Whatsup can update individual props precisely and pass atom references straight into DOM properties:

```tsx
function App() {
    const className = observable<'red' | 'green'>('red')

    const handleClick = () => {
        const newValue = className() === 'red' ? 'green' : 'red'
        className(newValue)
    }

    while (true) {
        yield(
            <button className={className} onClick={handleClick}>
                Click me
            </button>
        )
    }
}

render(<App />)
```

When `className` changes, the `while (true)` loop does **not** restart. There is no full re-render — only that DOM property updates. @brainkim wrote about this tradeoff above. I didn't choose observables because of this — they were already the foundation when I built the JSX layer — but this is exactly what they make possible.

Using a separate atom per DOM node sounds excessive and memory-hungry on paper. In practice, after a lot of optimization work, memory use ended up comparable to React + MobX, with significantly better performance on many workloads.

---

### Responding to @brainkim's points

**Generator state vs. observable state**

Agree — a generator component is already stateful. You can keep ordinary variables in the generator scope alongside observables:

```tsx
function* App() {
    const counter = observable(0)
    const increment = () => counter(counter() + 1)
    let i = 0

    while (true) {
        i++ // not tracked by an observable
        yield (
            <div>
                <p>You clicked {counter()} times</p>
                <button onClick={increment}>Click me</button>
            </div>
        )
    }
}
```

The observable layer isn't meant to replace all component state. It's the formal reactive graph that drives rendering and, crucially, enables fine-grained DOM updates. Local generator variables are fine for state that doesn't need to participate in that graph.

The advantage of observables is that reactive state is formally declared. The disadvantage is that you have to use observables. @brainkim put that well.

**Fine-grained prop updates**

Because each DOM node is its own atom, you can pass an observable directly into a prop and update only that prop — the generator loop doesn't re-run (see the `className` example above).

Crank has to diff the whole element. Whatsup trades a larger reactive surface for targeted DOM updates. Different tradeoffs — I think @brainkim's summary of that is fair.

**Props iterators vs. `yield` return values**

Crank's `for ({} of this)` pattern is elegant — props arrive as an iterator that closes on unmount, so cleanup in the generator body runs naturally without an explicit `finally`:

```tsx
function* Timer() {
    let seconds = 0
    const interval = setInterval(() => this.refresh(() => seconds++), 1000)
    for ({} of this) {
        yield <div>Seconds: {seconds}</div>
    }
    clearInterval(interval)
}
```

I considered that shape seriously. I went with props as the _return value_ of `yield` because it's the native way generators receive data — so I decided to leverage it for passing props.

**`try` / `catch` / `finally`**

But there's an important caveat: in Crank's model with props iterator, cleanup in the loop body looks elegant, but if an error occurs during execution, that cleanup isn't guaranteed as reliably as with `try` / `finally`.

For me, this was a key argument in favor of the `try` / `catch` / `finally` pattern: the `finally` block is guaranteed to run both on error and on unmount. This provides more reliable teardown and lets you build something like React's `ErrorBoundary` with `catch`.

**About `while (true)`**

Since a component needs to re-run when props change or when reactive dependencies change — we need an ordinary infinite loop, and `while (true)` serves that purpose logically. But, as @brainkim correctly noted, it looks annoying — ugly and confusing.

It would be nice to hide it behind something like `loop {}` or `render {}`, but JavaScript doesn't offer such constructs. We could do something like `for ({} of render){}` where `render` is an infinite iterator that can pass utility data and other useful information. But I didn't develop that idea just for the sake of it — I left `while (true)` as-is for better times.

It's important to keep in mind that components in Whatsup are ordinary `computed` atoms on generators. If we were to replace `while (true)` with something, it should be a construct that works equally well for components and for regular generator-based atoms alike.

**Props as observables**

@brainkim wrote:

> However, because Whatsup is based on observables, you can't have a props observable, because you can't yield from inside the observable because it is a callback-based API. Ultimately I think generators and props iterators and manual `refresh()` wins out in terms of ergonomics.

I'm not entirely sure what this means. Could you clarify?

---

### Where things stand

Development has been mostly paused for the last few years. I tried to get more people using it, without much success. In my experience, generators unfortunately put a lot of developers off. The experiment was interesting, but my enthusiasm ran out. Carrying an open-source project alone is hard.

Thank you for the interest. Happy to answer questions here or in the [Whatsup repo](https://github.com/whatsup/whatsup) if you want more details.
